fix #4675: adding an informer timeout handled directly in our client

shawkins opened the pull request:
## Description

Fix #4675

This seems like a little less work than updating the mock and other tests with fixed expectations.  Here we'll trigger the shutdown of the websocket after 5 - 10 minutes (the same as the go client).

Another thought is to eliminate the random strategy specifically for the mock client, but that also seems hackish.

## Type of change
<!---
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

<!--
Integration tests (https://github.com/fabric8io/kubernetes-client/tree/master/kubernetes-itests)
Please check integration tests and provide/improve tests if applicable.

Open your PR in Draft mode and verify all of the applicable Checklist items before marking your pull request as ready for review
-->

manusa commented:
In this PR we're killing the Watch client-wise as opposed to sending a parameter with a timeout. This is because our MockServer doesn't support parameters with variable values.

We'll probably want to follow up **in the future** with the improvement of the MockServer to allow for variable path parameters and then sending the timeout parameter in the URL for the Watch request including the jitter.

shawkins commented:
Addressed the conflict and also added a comment / refinement about the message processing behavior - what matters is how things are handled at the watchmanager.  We do not, and should not, try and guard against further message processing in the reflector.

shawkins commented:
After reviewing again decided this is best handled by an existing method on the http / websocket watch handlers - closeRequest is already supposed to trigger a graceful restart.  Doing this via the exception mechanism is not as good as it forces a re-list as well.

shawkins commented:
One, hopefully last, refinement which also is now setting the url timeout as well as a fallback.  This also highlights the scope of the hard-coded expectation issue.  There are close to 40 places where this appears and several of those are shared among multiple tests.

sunix commented on kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/AbstractWatchManager.java:
```diff
@@ -87,9 +97,20 @@
+   * attempt to reconnect
+   */
+  public void closeRequest() {
+    Optional.ofNullable(latestRequestState).ifPresent(s -> s.closed.set(true));
```
```suggestion
    Optional.ofNullable(latestRequestState).ifPresent(state -> state.closed.set(true));
```
